### PR TITLE
Switch to when, break statements

### PR DESCRIPTION
**Pull request by darthorimar**

### CODE REVIEW

File: j2k/newSrc/org/jetbrains/kotlin/j2k/NewCodeBuilder.kt [hunk 1]
```diff
@@ -468,28 +468,45 @@ class NewCodeBuilder {
...
+                ktWhenStatement.cases.forEach { it.accept(this) }
             }
         }
 
-        override fun visitDefaultSwitchCase(defaultSwitchCase: JKDefaultSwitchCase) {
-            printer.print("else ->")
-            printer.indented {
-                defaultSwitchCase.statements.forEach { it.accept(this) }
+        override fun visitKtWhenCase(ktWhenCase: JKKtWhenCase) {
+            printer.separated(", ", ktWhenCase.labels) {
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/NewCodeBuilder.kt]: renderList

File: j2k/newSrc/org/jetbrains/kotlin/j2k/NewCodeBuilder.kt [hunk 2]
```diff
@@ -468,28 +468,45 @@ class NewCodeBuilder {
...
+
+        private fun <T> Printer.separated(separator: String, items: List<T>, apply: (T) -> Unit) {
+            for ((index, item) in items.withIndex()) {
+                if (index > 0) this.printWithNoIndent(separator)
+                apply(item)
+            }
+        }
+
+        private fun Printer.blockOrStatement(statements: List<JKStatement>): Printer {
+            if (statements.size == 1)
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/NewCodeBuilder.kt]: `statements.singleOrNull()?.accept...`

File: j2k/newSrc/org/jetbrains/kotlin/j2k/tree/impl/j.kt [hunk 3]
```diff
@@ -201,4 +201,27 @@ class JKJavaAssignmentExpressionImpl(
...
+class JKJavaSwitchStatementImpl(
+    expression: JKExpression,
+    cases: List<JKJavaSwitchCase>
+) : JKJavaSwitchStatement, JKBranchElementBase() {
+    override var expression: JKExpression by child(expression)
+    override var cases: List<JKJavaSwitchCase> by children(cases)
+    override fun <R, D> accept(visitor: JKVisitor<R, D>, data: D): R = visitor.visitJavaSwitchStatement(this, data)
+}
+
+class JKJavaDefaultwitchCaseImpl(statements: List<JKStatement>) : JKJavaDefaultwitchCase, JKBranchElementBase() {
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/tree/impl/j.kt]: Switch? Typo

File: j2k/newSrc/org/jetbrains/kotlin/j2k/tree/j.kt [hunk 4]
```diff
@@ -86,4 +86,24 @@ interface JKJavaAssignmentExpression : JKExpression, JKBranchElement {
...
+    var expression: JKExpression
+    var cases: List<JKJavaSwitchCase>
+}
+
+interface JKJavaSwitchCase : JKTreeElement {
+    fun isDefault(): Boolean
+    var statements: List<JKStatement>
+}
+
+interface JKJavaDefaultwitchCase : JKJavaSwitchCase {
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/tree/j.kt]: Switch? Typo

File: j2k/newSrc/org/jetbrains/kotlin/j2k/conversions/SwitchStatementConversion.kt [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        }
+        defaultEntry?.let(result::add)
+        return result
+    }
+
+    private fun createWhenLabel(switchCase: JKJavaSwitchCase): JKKtWhenLabel =
+        when (switchCase) {
+            is JKJavaLabelSwitchCase ->
+                JKKtValueWhenLabelImpl(switchCase.label.also { it.detach(it.parent!!) })
+            else -> JKKtDefaultWhenLabelImpl()
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/conversions/SwitchStatementConversion.kt]: `JKKtElseWhenLabel`

File: j2k/newSrc/org/jetbrains/kotlin/j2k/conversions/SwitchStatementConversion.kt [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                block.block.statements = block.block.statements.filterNot(::isSwitchBreak)
+                return listOf(block)
+            }
+        }
+        return statementsToKeep
+    }
+
+    private fun convertCaseStatements(cases: List<JKJavaSwitchCase>, caseIndex: Int, allowBlock: Boolean = true): List<JKStatement> {
+        val case = cases[caseIndex]
+        val fallsThrough: Boolean = if (caseIndex == cases.lastIndex) {
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/conversions/SwitchStatementConversion.kt]: `: Boolean` is redundant here

File: j2k/newSrc/org/jetbrains/kotlin/j2k/conversions/SwitchStatementConversion.kt [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        for (statement in this) {
+            when(statement) {
+                is JKBreakStatement -> return false
+                is JKReturnStatement-> return false
+                //TODO add support of this when will be added
+//                is JKThrowStatement-> return false
+//                is JKContinueStatement-> return false
+            }
+            val psiStatement = context.backAnnotator(statement)
+            when (psiStatement) {
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/conversions/SwitchStatementConversion.kt]: `if (is 1, is 2 && psiS.canCompl...) return false`

File: j2k/newSrc/org/jetbrains/kotlin/j2k/conversions/SwitchStatementConversion.kt [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                JKKtValueWhenLabelImpl(switchCase.label.also { it.detach(it.parent!!) })
+            else -> JKKtDefaultWhenLabelImpl()
+        }
+
+    private fun convertCaseStatements(statements: List<JKStatement>, allowBlock: Boolean = true): List<JKStatement> {
+        val statementsToKeep = statements
+            .filterNot(::isSwitchBreak)
+            .map { it.copyTreeAndDetach() }
+
+        if (allowBlock && statementsToKeep.size == 1) {
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/conversions/SwitchStatementConversion.kt]: TODO: Investigate in which cases we should unwrap internal block

File: j2k/newSrc/org/jetbrains/kotlin/j2k/tree/impl/base.kt [hunk 9]
```diff
@@ -111,4 +114,35 @@ abstract class JKBranchElementBase : JKElementBase(), JKBranchElement {
...
     override val children: MutableList<Any> = mutableListOf()
+
+    override fun deepClone(): JKElementBase {
+        val cloned = super.deepClone() as JKBranchElementBase
+        val deepClonedChildren =
+            cloned.children.map {
+                when (it) {
+                    is JKElementBase -> it.deepClone()
+                    is List<*> -> (it as List<JKElementBase>).map { it.deepClone() }
+                    else -> TODO("Sth wrong")
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/tree/impl/base.kt]: `error` function, "Tree corrupted" message for ex

File: j2k/newSrc/org/jetbrains/kotlin/j2k/tree/impl/base.kt [hunk 10]
```diff
@@ -111,4 +114,35 @@ abstract class JKBranchElementBase : JKElementBase(), JKBranchElement {
...
+                }
+            }
+
+        deepClonedChildren.forEach { child ->
+            when (child) {
+                is JKElementBase -> {
+                    child.detach(child.parent!!)
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/tree/impl/base.kt]: `.detach(this@deepClone)` to run consistency checks

File: j2k/newSrc/org/jetbrains/kotlin/j2k/tree/impl/base.kt [hunk 11]
```diff
@@ -111,4 +114,35 @@ abstract class JKBranchElementBase : JKElementBase(), JKBranchElement {
...
+                    child.attach(cloned)
+                }
+                is List<*> -> (child as List<JKElementBase>).forEach {
+                    it.detach(it.parent!!)
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/tree/impl/base.kt]: `.detach(this@deepClone)` to run consistency checks

**Review by semoro — CHANGES_REQUESTED** [review]

File: j2k/newSrc/org/jetbrains/kotlin/j2k/conversions/SwitchStatementConversion.kt [hunk 12]
```diff
@@ -17,100 +17,84 @@ import org.jetbrains.kotlin.j2k.tree.impl.*
...
+                    if (it.size == 1 && cases.first().statements.singleOrNull() is JKBlockStatement)
+                        listOf(JKBlockStatementImpl(JKBlockImpl(it)))
+                    else it
+                }
+            val javaLabels = cases
+                .takeWhileInclusive { it.statements.isEmpty() }
+
+            val statementLabels = javaLabels
+                .filterIsInstance<JKJavaLabelSwitchCase>()
+                .map { JKKtValueWhenLabelImpl(it.label) }
```
Comment (semoro) [inline j2k/newSrc/org/jetbrains/kotlin/j2k/conversions/SwitchStatementConversion.kt:52]: Isn't it will cause tree corruption without `it.invalidate()` ?

Comment (darthorimar) [reply j2k/newSrc/org/jetbrains/kotlin/j2k/conversions/SwitchStatementConversion.kt:52]: No, because label was already detached on line 25

**Review by semoro — APPROVED** [review]

**Pull request merged**